libraries/GCS_MAVLink/GCS.h

```diff
@@ -107,6 +109,189 @@ enum ap_message : uint8_t {
+    void handle_mission_write_partial_list(GCS_MAVLINK &link,
+                                           const mavlink_message_t &msg,
+                                           const mavlink_mission_write_partial_list_t &packet);
+
+    // called on receipt of a MISSION_ITEM or MISSION_ITEM_INT packet;
+    // Mission_Command is used as an intermediate form for both of
+    // them:
+    void handle_mission_item(const mavlink_message_t &msg,
+                             AP_Mission::Mission_Command &cmd);
+
+    void handle_mission_clear_all(const GCS_MAVLINK &link,
+                                  const mavlink_message_t &msg);
+
+    void queued_request_send();
+    void update();
+
+    bool active_link_is(const GCS_MAVLINK *_link) const { return _link == link; };
+
+    virtual MAV_MISSION_TYPE mission_type() const = 0;
+
+    bool receiving; // currently sending requests and expecting items
+
+protected:
+
+    GCS_MAVLINK *link; // link currently receiving waypoints on
+
+    // return the ap_message which can be queued to be sent to send a
+    // item request to the GCS:
+    virtual ap_message next_item_ap_message_id() const = 0;
+
+    void handle_mission_clear_all();
+
+    virtual bool clear_all_items() = 0;
+
+private:
+
+    virtual void _handle_mission_count(class GCS_MAVLINK &link,
+                                       const mavlink_mission_count_t &packet,
+                                       const mavlink_message_t &msg) = 0;
+
+    uint16_t        request_i; // request index
+    uint16_t        request_last; // last request index
+
+    // waypoints
+    uint8_t        dest_sysid; // where to send requests
+    uint8_t        dest_compid; // "
+    uint32_t        timelast_receive_ms;
+    uint32_t        timelast_request_ms;
+    const uint16_t  upload_timeout_ms = 8000;
+
+    // support for GCS getting waypoints etc from us:
+    virtual MAV_MISSION_RESULT _handle_mission_request_int(const mavlink_mission_request_int_t &packet,
+                                                           mavlink_mission_item_int_t &ret_packet) = 0;
+
+    virtual MAV_MISSION_RESULT _handle_mission_request(
+        const GCS_MAVLINK &_link,
+        const mavlink_message_t &msg,
+        const mavlink_mission_request_t &packet,
+        mavlink_mission_item_t &ret_packet) = 0;
+
+    void init_send_requests(GCS_MAVLINK &_link,
+                            const mavlink_message_t &msg,
+                            const int16_t _request_first,
+                            const int16_t _request_last);
+
+    void send_mission_ack(const mavlink_message_t &msg, MAV_MISSION_RESULT result) const;
+    void send_mission_ack(const GCS_MAVLINK &link, const mavlink_message_t &msg, MAV_MISSION_RESULT result) const;
+
+    virtual uint16_t item_count() const = 0;
+    virtual uint16_t max_items() const = 0;
+
+    virtual MAV_MISSION_RESULT replace_cmd(const AP_Mission::Mission_Command &cmd) = 0;
```

Comment:
These shouldn't necessarily have to be `Mission_Command`'s why not allow the MAVLink representation? (And if it is a Mission_Command then how does rally get converted, I can't seem to find a sane path for it.

Reply:
The problem with this IMO is that rally points can't be represented safely in a mission command. You can't write a rally item to storage and have it round trip correctly. In practice we shouldn't be doing this, but it's making it easier to screw up by default. Each backend should be deciding how to do the conversion from the MAVLink message. Otherwise fences also have to be represented in AP_Mission...

Reply:
I had initially written this expecting to pass a misison-int to the subclasses.  I didn't find any code that was simply available to convert from a `MISSION_ITEM` to a `MISSION_ITEM_INT` and didn't want to pass both down to the backends.  The Mission_Command thing seems to work out - the fence support is an add-4-lines style of patch IIRC.

Reply:
... additionally, writing it with Mission_Cmd meant I didn't have a complete rewrite - I was reusing some code from the existing stuff.

I have patches locally now which move things to passing a MISSION_ITEM_INT - I'll clean them up a bit and push them up somewhere - if you think it's a much better idea then we can go that way.

Reply:
I've pushed these patches up here now.